### PR TITLE
[SPARK-35132][BUILD][CORE] Upgrade netty-all to 4.1.63.Final

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
There are 3 CVE problems were found after netty 4.1.51.Final as follows:

- [CVE-2021-21409](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-21409)
- [CVE-2021-21295](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-21295)
- [CVE-2021-21290](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-21290)

So the main change of this pr is upgrade netty-all to 4.1.63.Final avoid these potential risks.

Another change is to clean up deprecated api usage: [Tiny caches have been merged into small caches](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L447-L455)(after [netty#10267](https://github.com/netty/netty/pull/10267)) and [should use  PooledByteBufAllocator(boolean, int, int, int, int, int, int, boolean, int)](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledByteBufAllocator.java#L227-L239) api to create `PooledByteBufAllocator`.

### Why are the changes needed?
Upgrade netty-all to 4.1.63.Final avoid CVE problems.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass the Jenkins or GitHub Action

### CODE REVIEW

File: common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java [hunk 1]
```diff
@@ -162,7 +162,6 @@ public static PooledByteBufAllocator createPooledByteBufAllocator(
       Math.min(PooledByteBufAllocator.defaultNumDirectArena(), allowDirectBufs ? numCores : 0),
       PooledByteBufAllocator.defaultPageSize(),
       PooledByteBufAllocator.defaultMaxOrder(),
-      allowCache ? PooledByteBufAllocator.defaultTinyCacheSize() : 0,
```
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: Although there is a description about this, could you add the link to Netty site about this change?

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: done ~

Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: Thanks!

Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: Could you re-trigger your GitHub Action?
- https://github.com/LuciferYang/spark/actions/runs/762178571

**Review by dongjoon-hyun — APPROVED** [review]
LGTM (if CI passes).

cc @srowen

**Comment by LuciferYang** [comment]
image

It seems that there are some old netty jars in `deps/jars` directory

**Comment by LuciferYang** [comment]
I find that there is always netty 4.1.50 in classpath now, even for the master branch , this will cause some tests fail, and I created them https://issues.apache.org/jira/browse/SPARK-35134 to track this problem.

I try to execute  

```
mvn dependency:tree -Phadoop-3.2 -Phive-2.3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive
```
 there was no netty 4.1.50 dependency in the results.

Do you have any good suggestions? @dongjoon-hyun @HyukjinKwon @srowen

**Comment by HyukjinKwon** [comment]
Can we fix the dependency list?

```
To update the manifest file, run './dev/test-dependencies.sh --replace-manifest'.
diff --git a/dev/deps/spark-deps-hadoop-2.7-hive-2.3 b/dev/pr-deps/spark-deps-hadoop-2.7-hive-2.3
index 646aa49..8e25ad7 100644
--- a/dev/deps/spark-deps-hadoop-2.7-hive-2.3
+++ b/dev/pr-deps/spark-deps-hadoop-2.7-hive-2.3
@@ -191,7 +191,7 @@ metrics-jmx/4.1.1//metrics-jmx-4.1.1.jar
 metrics-json/4.1.1//metrics-json-4.1.1.jar
 metrics-jvm/4.1.1//metrics-jvm-4.1.1.jar
 minlog/1.3.0//minlog-1.3.0.jar
-netty-all/4.1.51.Final//netty-all-4.1.63.Final.jar
+netty-all/4.1.63.Final//netty-all-4.1.63.Final.jar
 objenesis/2.6//objenesis-2.6.jar
 okhttp/3.12.12//okhttp-3.12.12.jar
 okio/1.14.0//okio-1.14.0.jar
 ```

**Comment by LuciferYang** [comment]
@HyukjinKwon ok

**Comment by HyukjinKwon** [comment]
No, sorry @LuciferYang. Seems like the problemtic file was already fixed here.

**Comment by HyukjinKwon** [comment]
It's weird that it fails ..

**Comment by HyukjinKwon** [comment]
I think there's a dependency resolution diff between Maven and SBT. You might have to manually exclude or override the version in `SparkBuild.scala` e.g.) `ExcludeDependencides` or `DependencyOverrides`

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3 [hunk 2]
```diff
@@ -191,7 +191,7 @@ metrics-jmx/4.1.1//metrics-jmx-4.1.1.jar
 metrics-json/4.1.1//metrics-json-4.1.1.jar
 metrics-jvm/4.1.1//metrics-jvm-4.1.1.jar
 minlog/1.3.0//minlog-1.3.0.jar
-netty-all/4.1.51.Final//netty-all-4.1.51.Final.jar
+netty-all/4.1.51.Final//netty-all-4.1.63.Final.jar
```
Comment (HyukjinKwon) [inline dev/deps/spark-deps-hadoop-2.7-hive-2.3:194]: Oh, yeah, and seems like we should do `./dev/test-dependencies.sh --replace-manifest`:

```diff
-netty-all/4.1.51.Final//netty-all-4.1.63.Final.jar
+netty-all/4.1.63.Final//netty-all-4.1.63.Final.jar
```

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3]: `./dev/test-dependencies.sh --replace-manifest:` 
thx @HyukjinKwon ~ learned this new skill ~

**Comment by LuciferYang** [comment]
> I think there's a dependency resolution diff between Maven and SBT. You might have to manually exclude or override the version in SparkBuild.scala e.g.) ExcludeDependencides or DependencyOverrides

Let me try this

**Comment by LuciferYang** [comment]
> I think there's a dependency resolution diff between Maven and SBT. You might have to manually exclude or override the version in SparkBuild.scala e.g.) ExcludeDependencides or DependencyOverrides

https://github.com/apache/spark/pull/32230 try to solve this problem

File: common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java (hunk 1 above)
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: ok

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java:165]: There are some netty 4.1.50 dependencies in sbt test classpath, I'm trying to add some `ExcludeRule` to fix it now.

**Comment by LuciferYang** [comment]
cherry-pick 3e93efe from SPARK-35134 to check GA test result.

**Comment by LuciferYang** [comment]
It seems that the fix of https://github.com/apache/spark/pull/32230# can solve the netty version conflict in the test

**Comment by LuciferYang** [comment]
@dongjoon-hyun @HyukjinKwon @srowen All test passed now

**Review by HyukjinKwon — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by LuciferYang** [comment]
thx @HyukjinKwon @dongjoon-hyun @srowen

**Pull request closed without merging**